### PR TITLE
Support tri-color waveshare eink displays 2.7inch B and B V2

### PR DESCRIPTION
**Pull request by rnauber**
# What does this implement/fix?

This patch provides the infrastructure to support tri-color (black, white, red) displays and implements drivers for the "2.7inch B" variants.

photo_2022-12-29_19-08-03

  
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation:** esphome/esphome-docs#3580

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
```yaml

spi:
  clk_pin: 14
  mosi_pin: 27

display:
  - platform: waveshare_epaper
    cs_pin: 32
    dc_pin: 25
    busy_pin: 33
    reset_pin: 26
    model: 2.70in-b
    reset_duration: 2ms
    update_interval: 60s
    lambda: |-
      const auto RED = Color(0xff,0,0,0);
      const auto BLACK = Color(0,0,0,0);
      const auto WHITE = Color(0xff,0xff,0xff,0xff);
      it.fill(WHITE);
           
      it.print(10, 0, id(myfont), BLACK, "Hello");
      it.print(10, 40, id(myfont), RED, "World!");
  
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by rnauber** [comment]
> There hasn't been any activity on this pull request recently. This pull request has been automatically marked as stale because of that and will be closed if no further activity occurs within 7 days. Thank you for your contributions.

Hi everyone,
is there some interest in merging this patch, @jesserockz ? Which changes would be needed? Otherwise, I am fine in just using it on my own hardware :)

Best,
Richard

**Comment by lboue** [comment]
@rnauber
Is it possible to reopen this PR because it was not accepted?

**Comment by rnauber** [comment]
Hi everyone,
I synced this PR to the current dev branch, it works on my device and passes all the tests.  @jesserockz , is there interest in merging this? Everyone can test this PR with something like this:

```
external_components:
  - source:
      type: git
      url: https://github.com/rnauber/esphome.git
      ref: feature/waveshare_eink_color
    refresh: 1min
    components: [ waveshare_epaper ]

   
color:
  - id: my_red
    red: 100%
    green: 0%
    blue: 0%
    white: 0%      
  - id: my_black
    red: 0%
    green: 0%
    blue: 0%
    white: 0%    
  - id: my_white
    red: 0%
    green: 0%
    blue: 0%
    white: 100%     

display:
  - platform: waveshare_epaper
    cs_pin: 32
    dc_pin: 25
    busy_pin: 33
    reset_pin: 26
    model: 2.70in-b
    reset_duration: 2ms
    update_interval: 60s
    rotation: 90°
    lambda: |-
      it.fill(id(my_white));
```

Best,
Richard

**Comment by darianndd** [comment]
Hi,

Thank you for your effort to make Red/Black/White e-ink displays work!
I really like Esphome project and I think this is a good start to have RBW einks supported.

However, I have used your change to run with my RBW eink, but it seems that Waveshare products works differently while they should only differ in resolution. I did a quick, but dirty adaptation of resolution in WaveshareEPaper2P7InBV2 class, but without success. The display background turned black, "World!" text was in red, but "Hello!" wasn't visible at all. On the top of display was thin area with random black and red dots which suggested that printed buffer is not in valid position. To make it work, I had to add "update control" code and use loops from WaveshareEPaper2P7InB::display(), thus I got a mix of two your classes (diff below) and then it started working... weird. In the end, Esphome will have a set of classes dedicated to specific displays... it seems that there is no common (or at least very little) code to handle as many displays as possible, so I think it is better to have more classes (as you added) instead of many conditions in generic methods.

Will this patch be merged soon?

```
diff --git a/esphome/components/waveshare_epaper/waveshare_epaper.cpp b/esphome/components/waveshare_epaper/waveshare_epaper.cpp
index bb41269a..fb8f7c47 100644
--- a/esphome/components/waveshare_epaper/waveshare_epaper.cpp
+++ b/esphome/components/waveshare_epaper/waveshare_epaper.cpp
@@ -1028,6 +1028,11 @@ void WaveshareEPaper2P7InBV2::initialize() {
   this->data(yend & 0xff);
   this->data((yend >> 8) & 0xff);
 
+  // Display update control - to properly set the view
+  this->command(0x21);
+  this->data(0x00);
+  this->data(0x80);
+
   // SetCursor(self, Xstart, Ystart):
   this->command(0x4E);
   this->data(0x00);
@@ -1038,14 +1043,16 @@ void WaveshareEPaper2P7InBV2::initialize() {
   ESP_LOGI(TAG, "Done Eink init");
 }
 void HOT WaveshareEPaper2P7InBV2::display() {
-  uint32_t buf_len = this->get_buffer_length_();
+  uint32_t buf_len_half = this->get_buffer_length_() >> 1;
 
   ESP_LOGI(TAG, "Refresh ... ");
 
+  this->wait_until_idle_();
+
   // COMMAND DATA START TRANSMISSION 1 (BLACK)
   this->command(0x24);
   delay(2);
-  for (uint32_t i = 0; i < buf_len; i++) {
+  for (uint32_t i = 0; i < buf_len_half; i++) {
     this->data(this->buffer_[i]);
   }
   delay(2);
@@ -1053,7 +1060,7 @@ void HOT WaveshareEPaper2P7InBV2::display() {
   // COMMAND DATA START TRANSMISSION 2  (RED)
   this->command(0x26);
   delay(2);
-  for (uint32_t i = 0; i < buf_len; i++) {
+  for (uint32_t i = buf_len_half; i < buf_len_half * 2u; i++) {
     this->data(this->buffer_[i]);
   }
 
@@ -1067,8 +1074,8 @@ void HOT WaveshareEPaper2P7InBV2::display() {
 
   ESP_LOGI(TAG, "... done ");
 }
-int WaveshareEPaper2P7InBV2::get_width_internal() { return 176; }
-int WaveshareEPaper2P7InBV2::get_height_internal() { return 264; }
+int WaveshareEPaper2P7InBV2::get_width_internal() { return 152; }
+int WaveshareEPaper2P7InBV2::get_height_internal() { return 296; }
 void WaveshareEPaper2P7InBV2::dump_config() {
   LOG_DISPLAY("", "Waveshare E-Paper", this);
   ESP_LOGCONFIG(TAG, "  Model: 2.7in B V2");
```
I used your yaml file from comment above, but with my own "display" section:

```
display:                                                                                                                                                                                                                                                                                  
  - platform: waveshare_epaper                                                                                                                                                                                                                                                            
    id: disp1                                                                                                                                                                                                                                                                             
    cs_pin: GPIO14                                                                                                                                                                                                                                                                        
    dc_pin: GPIO27                                                                                                                                                                                                                                                                        
    busy_pin: GPIO13                                                                                                                                                                                                                                                                      
    reset_pin: GPIO26                                                                                                                                                                                                                                                                     
    model: 2.70in-bv2                                                                                                                                                                                                                                                                     
    rotation: 90°                                                                                                                                                                                                                                                                         
    reset_duration: 2ms                                                                                                                                                                                                                                                                   
    update_interval: 60s                                                                                                                                                                                                                                                                  
    lambda: |-                                                                                                                                                                                                                                                                            
      const auto RED = Color(0xff, 0, 0, 0);                                                                                                                                                                                                                                              
      const auto BLACK = Color(0, 0, 0, 0);                                                                                                                                                                                                                                               
      const auto WHITE = Color(0xff, 0xff, 0xff, 0xff);                                                                                                                                                                                                                                   
      it.fill(WHITE);                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                          
      static int i = 0;                                                                                                                                                                                                                                                                   
      it.printf(0, 0, id(font_vcr_21), BLACK, "Hello!");                                                                                                                                                                                                                                  
      it.printf(0, 30, id(font_vcr_21), RED, "World!");                                                                                                                                                                                                                                   
      it.printf(0, 60, id(font_vcr_21), RED, "v: %d", i++);                                                                                                                                                                                                                               
```

![einks_front](https://github.com/esphome/esphome/assets/23472395/2effa2d0-37ec-45ad-a9fb-4b5b9adeb3b0)
 
![einks_back](https://github.com/esphome/esphome/assets/23472395/97e2d34f-2eb8-46c9-96bc-9b59b80d927a)

**Review by jesserockz — DISMISSED** [review]
Please rebase your changes

**Comment by rnauber** [comment]
> Please rebase your changes

Hi @jesserockz , I rebased the patch, made it pass the tests and tested it on my device!

Best,
Richard

File: esphome/components/waveshare_epaper/display.py [hunk 1]
```diff
@@ -20,62 +20,70 @@
 WaveshareEPaper = waveshare_epaper_ns.class_(
     "WaveshareEPaper", cg.PollingComponent, spi.SPIDevice, display.DisplayBuffer
 )
+WaveshareEPaperBW = waveshare_epaper_ns.class_("WaveshareEPaperBW", WaveshareEPaper)
+WaveshareEPaperBWR = waveshare_epaper_ns.class_("WaveshareEPaperBWR", WaveshareEPaper)
 WaveshareEPaperTypeA = waveshare_epaper_ns.class_(
-    "WaveshareEPaperTypeA", WaveshareEPaper
```
Comment (clydebarrow) [inline esphome/components/waveshare_epaper/display.py:24]: Was it necessary to rename this class? While you might have wanted it to be more descriptive, it adds a lot of clutter to the PR, making it harder to review.

File: esphome/components/waveshare_epaper/waveshare_epaper.h [hunk 2]
```diff
@@ -56,7 +50,7 @@ class WaveshareEPaper : public display::DisplayBuffer,
 
   virtual int get_width_controller() { return this->get_width_internal(); };
 
-  uint32_t get_buffer_length_();
+  virtual uint32_t get_buffer_length_() = 0;  // NOLINT
```
Comment (clydebarrow) [inline esphome/components/waveshare_epaper/waveshare_epaper.h]: Why the NOLINT here?

**Review by clydebarrow — CHANGES_REQUESTED** [review]

**Comment by clydebarrow** [comment]
Is there a corresponding doc PR?

File: esphome/components/waveshare_epaper/display.py (hunk 1 above)
Comment (clydebarrow) [reply esphome/components/waveshare_epaper/display.py:24]: An alternative would be to create a new class e.g. `WaveshareEPaperBase` to act as the superclass for the BW (with the original name) and BWR classes. That would minimise the textual changes, which not only makes review easier, but keeps the history cleaner.

Comment (rnauber) [reply esphome/components/waveshare_epaper/display.py:24]: Hi @clydebarrow , thank you for your swift review! 
I am ok with having  `WaveshareEPaperBase` as superclass and  `WaveshareEPaper` and  `WaveshareEPaperBWR` as children. I will refactor it and get back to you...
Thanks again!
Richard

File: esphome/components/waveshare_epaper/waveshare_epaper.h (hunk 2 above)
Comment (rnauber) [reply esphome/components/waveshare_epaper/waveshare_epaper.h]: Ah, good catch, it was useless! I removed it...

File: esphome/components/waveshare_epaper/display.py (hunk 1 above)
Comment (rnauber) [reply esphome/components/waveshare_epaper/display.py:24]: Ok, I refactored it.

**Comment by rnauber** [comment]
> Is there a corresponding doc PR?
Now there is: 
https://github.com/esphome/esphome-docs/pull/3580

File: esphome/components/waveshare_epaper/waveshare_epaper.h (hunk 2 above)
Comment (rnauber) [reply esphome/components/waveshare_epaper/waveshare_epaper.h]: I stand corrected, the linter will complain about  "get_buffer_length_()", if there is no NOLINT

**Review by clydebarrow — APPROVED** [review]
LGTM, has not broken the B&W versions, so that's good!

**Comment by rnauber** [comment]
Thanks for reviewing and merging this PR, @clydebarrow !

**Pull request merged**